File: 735-asteroid-collision/asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        ret = []
        for ast in asteroids:
            while ret and ret[-1]>0 and ast<0:
                if abs(ast)>ret[-1]:
                    ret.pop(-1)
                elif abs(ast)==ret[-1]:
                    ret.pop(-1)
                    break
                else:
                    break
            else:
                ret.append(ast)
        return ret

File: 735-asteroid-collision/asteroid-collision.ts
function asteroidCollision(asteroids: number[]): number[] {
    const stack: number[] = [];
    
    for (const ast of asteroids) {
        let survived = true;
        
        while (stack.length > 0 && stack[stack.length - 1] > 0 && ast < 0) {
            const top = stack[stack.length - 1];
            if (top < -ast) {
                stack.pop(); // top destroyed, continue checking
                continue;
            } else if (top === -ast) {
                stack.pop(); // both destroyed
            }
            // if top > -ast or equal, current asteroid is destroyed
            survived = false;
            break;
        }
        
        if (survived) {
            stack.push(ast);
        }
    }
    
    return stack;
}